File: frontend/utils/update-env.js
const os = require('os');
const fs = require('fs');
const path = require('path');


const EXCLUDED_INTERFACE_NAME_RE = /vEthernet|WSL|Hyper-V|Loopback|docker|vboxnet|vmnet|virbr/i;
const PREFERRED_LAN_IP_RE = /^(10\.|192\.168\.)/;

function isWslEnvironment() {
  if (process.env.WSL_DISTRO_NAME || process.env.WSL_INTEROP) return true;
  try {
    return fs.readFileSync('/proc/version', 'utf8').toLowerCase().includes('microsoft');
  } catch {
    return false;
  }
}

function getPreferredIP() {
  try {
    const interfaces = os.networkInterfaces();
    if (!interfaces || typeof interfaces !== 'object') return null;

    for (const name of Object.keys(interfaces)) {
      if (EXCLUDED_INTERFACE_NAME_RE.test(name)) continue;
      for (const iface of interfaces[name] || []) {
        if (
          iface.family === 'IPv4' &&
          !iface.internal &&
          PREFERRED_LAN_IP_RE.test(iface.address)
        ) {
          return iface.address;
        }
      }
    }
  } catch (err) {
    console.warn(`⚠️ IP detection failed (${err.message}). Falling back to localhost.`);
  }

  return null;
}

const explicitHost = (process.env.BACKEND_HOST || '').trim();
const autoDetect = /^(1|true|yes)$/i.test((process.env.AUTO_DETECT_BACKEND_IP || '').trim());
const runningInWsl = isWslEnvironment();
const detectedIp = autoDetect ? getPreferredIP() : null;
const ip = explicitHost || (runningInWsl ? 'localhost' : (detectedIp || 'localhost'));
const envPath = path.join(__dirname, '../.env.local');
function upsertEnvVars(envFilePath, updates) {
  const existing = fs.existsSync(envFilePath)
    ? fs.readFileSync(envFilePath, 'utf8')
    : '';
  const lines = existing.split(/\r?\n/);
  const seen = new Set();
  const output = [];

  for (const line of lines) {
    const match = line.match(/^\s*([A-Za-z_][A-Za-z0-9_]*)\s*=/);
    if (!match) {
      if (line.length) output.push(line);
      continue;
    }
    const key = match[1];
    if (Object.prototype.hasOwnProperty.call(updates, key)) {
      output.push(`${key}=${updates[key]}`);
      seen.add(key);
    } else {
      output.push(line);
    }
  }

  for (const [key, value] of Object.entries(updates)) {
    if (!seen.has(key)) output.push(`${key}=${value}`);
  }

  const nextContent = `${output.filter(Boolean).join('\n')}\n`;
  fs.writeFileSync(envFilePath, nextContent, 'utf8');
  return nextContent;
}

const content = upsertEnvVars(envPath, {
  NEXT_PUBLIC_API_BASE_URL: `http://${ip}:8000`,
  NEXT_PUBLIC_WS_URL: `ws://${ip}:8000/ws/translate`,
});

if (ip === 'localhost') {
  const wslHint = runningInWsl && !explicitHost
    ? '\nWSL detected: using localhost avoids unstable WSL private IPs like 172.x.x.x.'
    : '';
  console.log(`⚠️ .env.local updated with localhost default:\n${content}${wslHint}`);
} else {
  console.log(`✅ .env.local updated with backend host (${ip}):\n${content}`);
}

File: update_env.py
import socket
import os
from pathlib import Path


def get_local_ip():
    s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
    try:
        s.connect(("10.255.255.255", 1))
        return s.getsockname()[0]
    except Exception:
        return "127.0.0.1"
    finally:
        s.close()


def upsert_env_vars(path: Path, updates: dict[str, str]) -> str:
    existing = path.read_text(encoding="utf-8") if path.exists() else ""
    lines = existing.splitlines()
    seen: set[str] = set()
    output: list[str] = []

    for line in lines:
        stripped = line.strip()
        if not stripped or "=" not in line:
            if line:
                output.append(line)
            continue
        key = line.split("=", 1)[0].strip()
        if key in updates:
            output.append(f"{key}={updates[key]}")
            seen.add(key)
        else:
            output.append(line)

    for key, value in updates.items():
        if key not in seen:
            output.append(f"{key}={value}")

    next_content = "\n".join([line for line in output if line]) + "\n"
    path.write_text(next_content, encoding="utf-8")
    return next_content


explicit_host = (os.getenv("BACKEND_HOST") or "").strip()
auto_detect = (os.getenv("AUTO_DETECT_BACKEND_IP") or "").strip().lower() in {"1", "true", "yes"}
ip = explicit_host or (get_local_ip() if auto_detect else "localhost")

frontend_env_path = Path("frontend") / ".env.local"
frontend_env_path.parent.mkdir(parents=True, exist_ok=True)
env_text = upsert_env_vars(
    frontend_env_path,
    {
        "NEXT_PUBLIC_API_BASE_URL": f"http://{ip}:8000",
        "NEXT_PUBLIC_WS_URL": f"ws://{ip}:8000",
    },
)

print("✅ .env.local updated with:")
print(env_text)
